### PR TITLE
Parser: rewrite tuple expr recovery to allow better items recovery

### PR DESCRIPTION
**Pull request by auduchinok**
Rewrites tuple expression recovery in a way that allows having better recovery for tuple item expressions. It also changes recovery for `=` binary expressions, which allows a significantly better recovery for named arguments analysis (and improved Parameter Info feature quality), if the type checker is also updated later.

Due to how precedences between tuple and binary expressions are aligned in F# (and how there's no separate parser rules for things like argument lists), it's difficult to have great recovery for both at the same time. This PR worsens the tuple recovery a bit, by removing rules for missing first item in the most cases:
```fsharp
let _ = , a
```
but adds a special rule to still allow that recovery inside parens, e.g. in a method call:
```fsharp
obj.Method(, b)
obj.Method(, b, c)
```
This rule implementation is more ad-hoc than other rules, but it may be a good compromise in this particular situation.
Recovery for missing items between and after other items is unchanged and is still good.

At the same time it adds recovery for previously failed cases like this:
```fsharp
M(a = 1, b = )
M(a =  , b = 2)
```

The balance seems to be good, since it's probably more common to write an additional named arg in an existing argument list, than to prepend an expression with a comma and then with another expression.

This PR also makes tuple expr recovery tests use the newer tree dumps checks instead of manually checking parts of the tree.

If this PR works out well, I'll likely try to expand the change into tuple patterns and other binary expressions, as it currently covers only `=` expressions and tuple expressions to make it more manageable.

### CODE REVIEW

**Comment by T-Gro** [comment]
What do we now get in the "regressed" scenario? I did not see it in the tests (or missed it)

``` let _ = , a ```

File: src/Compiler/pars.fsy [hunk 1]
```diff
@@ -4370,23 +4366,47 @@ tupleExpr:
...
+        if not $4 then reportParseErrorAt mComma2 (FSComp.SR.parsExpectedExpressionAfterToken ())
+        let expr1 = arbExpr ("tupleExpr6", mComma1.EndRange)
+        let expr2 = arbExpr ("tupleExpr7", mComma2.EndRange)
+        [expr2; expr1; $1], [mComma2; mComma1] }
+
+  | declExpr COMMA COMMA declExpr
+      { let mComma1 = rhs parseState 2
+        let mComma2 = rhs parseState 3
+        reportParseErrorAt mComma2 (FSComp.SR.parsExpectingExpressionInTuple ())
+        let expr = arbExpr ("tupleExpr8", mComma1.EndRange)
```
[the remark this answers is not in the document]
Comment (auduchinok) [reply src/Compiler/pars.fsy:4408]: Yes, these strings are stored in produced `ArbitraryAfterError` nodes, and it helps tremendously when you're trying to analyze the tree and want to see what particular recovery rule was used. These strings are seen in the test baselines, debugger, and in [Fantomas tools](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWQ2wAuYfAXjABGXLgDMkEAF8gA).

**Comment by auduchinok** [comment]
> What do we now get in the "regressed" scenario? I did not see it in the tests (or missed it)

It's covered here:
https://github.com/dotnet/fsharp/pull/15227/files#diff-defca677f09ea14864047eefb5081844627ef68e3e4216daefb17079197fadb5

Not a good output, but I consider it to be a much less critical case than the ones that got working now, like these:

```fsharp
M(a = 1, b = , c = 3)
M(a =, b = 2)
```

The removed rule was also made by me, and now I think it wasn't the best or an entirely correct way to do it, despite fixing the parsing. Why incorrect? Because its precedence is too high, and it would try to parse things as tuples too aggressively in cases where it could be better to have recovery in the outer nodes that have lower precedence.

The problem with the outer nodes recovery was that in this case:

```fsharp
a =, b = 2 // from M(a =, b = 2)
```

it would be parsed as
```
binaryApp
  a
  =
  tuple
    error
    binaryApp
      b
      =
      2
```
instead of more expected
```
tuple
  binaryApp
    a
    =
    error
  binaryApp
    b
    =
    2
```

This is the main kind of cases that this PR fixes.

**Review by T-Gro — APPROVED** [review]

**Comment by T-Gro** [comment]
That makes sense, thanks.
Approved now.

**Review by edgarfgp — APPROVED** [review]

**Comment by auduchinok** [comment]
This is ready.

**Pull request merged**